src/main/scala/decider/TermToZ3APIConverter.scala

```diff
@@ -0,0 +1,525 @@
+      case m: MapUpdate => createApp("Map_update", Seq(m.base, m.key, m.value), m.sort)
+
+      /* Quantified Permissions */
+
+      case Domain(id, fvf) => createApp("$FVF.domain_" + id, Seq(fvf), term.sort)
+
+      case Lookup(field, fvf, at) =>
+        createApp("$FVF.lookup_" + field, Seq(fvf, at), term.sort)
+
+      case FieldTrigger(field, fvf, at) => createApp("$FVF.loc_" + field, (fvf.sort match {
+        case sorts.FieldValueFunction(_) => Seq(Lookup(field, fvf, at), at)
+        case _ => Seq(fvf, at)
+      }), term.sort)
+
+      case PermLookup(field, pm, at) => createApp("$FVF.perm_" + field, Seq(pm, at), term.sort)
+
+      case PredicateDomain(id, psf) => createApp("$PSF.domain_" + id, Seq(psf), term.sort)
+
+      case PredicateLookup(id, psf, args) =>
+        val snap: Term = toSnapTree(args)
+        createApp("$PSF.lookup_" + id, Seq(psf, snap), term.sort)
+
+      case PredicateTrigger(id, psf, args) =>
+        val snap: Term = toSnapTree(args)
+        createApp("$PSF.loc_" + id, Seq(PredicateLookup(id, psf, args), snap), term.sort)
+
+      case PredicatePermLookup(predname, pm, args) =>
+        val snap: Term = toSnapTree(args)
+        createApp("$PSF.perm_" + predname, Seq(pm, snap), term.sort)
+
+      /* Other terms */
+
+      case First(t) => ctx.mkApp(firstFunc, convertTerm(t))
+      case Second(t) => ctx.mkApp(secondFunc, convertTerm(t))
+
+      case bop: Combine =>
+        ctx.mkApp(combineConstructor, convertTerm(bop.p0), convertTerm(bop.p1))
+
+      case SortWrapper(t, to) =>
+        createApp(convertId(SortWrapperId(t.sort, to)), Seq(t), to)
+
+      case Distinct(symbols) =>
+        ctx.mkDistinct(symbols.map(s => ctx.mkConst(convertId(s.id), convertSort(s.resultSort))).toSeq: _*)
+
+      case Let(bindings, body) =>
+        convert(body.replace(bindings))
+
+      case _: MagicWandChunkTerm
+         | _: Quantification =>
+        sys.error(s"Unexpected term $term cannot be translated to SMTLib code")
+    }
+    res
+  }
+
+  @inline
+  protected def createApp(functionName: String, args: Seq[Term], outSort: Sort): Z3Expr = {
+    ctx.mkApp(getFuncDecl(functionName, outSort, args.map(_.sort)), args.map(convertTerm(_)): _*)
+  }
+
+  def getFuncDecl(name: String, resSort: Sort, argSorts: Seq[Sort]): Z3FuncDecl = {
+    val existingEntry = funcDeclCache.get((name, argSorts, resSort))
+    if (existingEntry.isDefined)
+      return existingEntry.get
+    val res = ctx.mkFuncDecl(name, argSorts.map(a => convertSort(a)).toArray, convertSort(resSort))
+    funcDeclCache.update((name, argSorts, resSort), res)
+    res
+  }
+
+
+  @inline
+  protected def createSMTApp(functionName: String, args: Seq[Term], outSort: Sort): Z3Expr = {
+    // workaround: since we cannot create a function application with just the name, we let Z3 parse
+    // a string that uses the function, take the AST, and get the func decl from there, so that we can
+    // programmatically create a func app.
```

Comment:
Is this because we don't know the types for the arguments? Are we passing them to Z3 so it can figure out their sorts?

Reply:
No, the problem is that there is nothing like ctx.getInterpretedFuncDecl(name) in the API that I could give a string like "fp.add" to and get the declaration for floating point addition. Instead, for any built-in interpreted function, there is a separate creator method in the API (e.g., ctx.mkFloatAdd(arg1, arg2)). So I'd have to have a big match-case here that matches all names of all built-in functions and then calls the respectice API method. Which I didn't want to do now because there are a lot of them, and the way we currently use SMTFuncApps, sometimes, the strings aren't actually *just* a function name but a function name and some number of default argument (e.g., "fp.add RNE", which contains the function name and a first argument that determines the rounding mode).